query/glob.c

```diff
@@ -0,0 +1,459 @@
+
+  uint32_t num_children_allocd;
+};
+
+// Look ahead in pattern; we want to find the directory separator.
+// While we are looking, check for wildmatch special characters.
+// If we do not find a directory separator, return NULL.
+static inline const char *find_sep_and_specials(const char *pattern,
+                                                const char *end,
+                                                bool *had_specials) {
+  *had_specials = false;
+  while (pattern < end) {
+    switch (*pattern) {
+      case '*':
+      case '?':
+      case '[':
+      case '\\':
+        *had_specials = true;
+        break;
+      case '/':
+        return pattern;
+    }
+    ++pattern;
+  }
+  // No separator found
+  return NULL;
+}
+
+void free_glob_tree(struct watchman_glob_tree *node) {
+  uint32_t i;
+
+  if (!node) {
+    return;
+  }
+
+  for (i = 0; i < node->num_children; ++i) {
+    free_glob_tree(node->children[i]);
+  }
+  free(node->children);
+  free(node);
+}
+
+static struct watchman_glob_tree *make_node(const char *pattern,
+                                            uint32_t pattern_len) {
+  struct watchman_glob_tree *node;
+
+  node = calloc(1, sizeof(*node) + pattern_len + 1);
+  if (!node) {
+    return NULL;
+  }
+
+  node->pattern = (char*)(node + 1);
+  memcpy(node->pattern, pattern, pattern_len);
+  node->pattern[pattern_len] = '\0';
+
+  node->pattern_len = pattern_len;
+
+  return node;
+}
+
+// Simple brute force lookup of pattern within a node.
+// This is run at compile time and most glob sets are low enough cardinality
+// that this doesn't turn out to be a hot spot in practice.
+static struct watchman_glob_tree *
+lookup_node_child(struct watchman_glob_tree *parent, const char *pattern,
+                  uint32_t pattern_len) {
+  uint32_t i;
+
+  for (i = 0; i < parent->num_children; ++i) {
+    if (parent->children[i]->pattern_len == pattern_len &&
+        memcmp(parent->children[i]->pattern, pattern, pattern_len) == 0) {
+      return parent->children[i];
+    }
+  }
+  return NULL;
+}
+
+// Add a child node to parent, allocating more space if needed.
+static bool add_node(struct watchman_glob_tree *parent,
+                     struct watchman_glob_tree *child) {
+  if (!parent->children) {
+    parent->num_children_allocd = 8;
+    parent->children =
+        calloc(parent->num_children_allocd, sizeof(*parent->children));
+    if (!parent->children) {
+      return false;
+    }
+  } else if (parent->num_children + 1 > parent->num_children_allocd) {
+    struct watchman_glob_tree **bigger =
+        realloc(parent->children,
+                parent->num_children_allocd * 2 * sizeof(*parent->children));
+    if (!bigger) {
+      return false;
+    }
+    parent->num_children_allocd *= 2;
+    parent->children = bigger;
+  }
+
+  parent->children[parent->num_children++] = child;
+  return true;
+}
+
+// Compile and add a new glob pattern to the tree.
```

Comment:
Cool. If we could normalize here, we could add the ability to compare two globs to handle the duplicate glob situation I mentioned.